pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,573 @@
+            provided, and the gradient recipe contains an unshifted term, this value is used,
+            saving a quantum evaluation.
+        validate_params (bool): Whether to validate the tape parameters or not. If ``True``,
+            the ``Operation.grad_method`` attribute and the circuit structure will be analyzed
+            to determine if the trainable parameters support the finite-difference method,
+            inferring that they support SPSA as well.
+            If ``False``, the SPSA gradient method will be applied to all parameters without
+            checking.
+        shots (None, int, list[int], list[ShotTuple]): The device shots that will be used to
+            execute the tapes outputted by this transform. Note that this argument doesn't
+            influence the shots used for tape execution, but provides information
+            to the transform about the device shots and helps in determining if a shot
+            sequence was used to define the device shots for the new return types output system.
+        num_directions (int): Number of sampled simultaneous perturbation vectors. An estimate for
+            the gradient is computed for each vector using the underlying finite-difference
+            method, and afterwards all estimates are averaged.
+
+    Returns:
+        tensor_like or tuple[tensor_like] or tuple[tuple[tensor_like]] or tuple[list[QuantumTape], function]:
+
+        - If the input is a QNode, an object representing the output Jacobian matrix.
+          The type of the object returned is either a tensor, a tuple or a nested tuple depending on the nesting
+          structure of the output.
+
+        - If the input is a tape, a tuple containing a list of generated tapes,
+          in addition to a post-processing function to be applied to the
+          evaluated tapes.
+
+    **Example**
+
+    This gradient transform can be applied directly to :class:`QNode <pennylane.QNode>`
+    objects:
+
+    >>> @qml.qnode(dev)
+    ... def circuit(params):
+    ...     qml.RX(params[0], wires=0)
+    ...     qml.RY(params[1], wires=0)
+    ...     qml.RX(params[2], wires=0)
+    ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+    >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+    >>> qml.gradients.spsa_grad(circuit)(params)
+    ((tensor(-0.19280803, requires_grad=True),
+      tensor(-0.19280803, requires_grad=True),
+      tensor(0.19280803, requires_grad=True)),
+     (tensor(0.34786926, requires_grad=True),
+      tensor(0.34786926, requires_grad=True),
+      tensor(-0.34786926, requires_grad=True)))
+
+    Note that the SPSA gradient is a statistical estimator that uses a given number of
+    function evaluations that does not depend on the number of parameters. While this
+    bounds the cost of the estimation, it also implies that the returned values are not
+    exact (even for devices with ``shots=None``) and that they will fluctuate.
```

Comment:
Should we provide a way to fix the randomness?

Reply:
I think a `seed` keyword argument makes sense here, then we can pass it to the sampler. Do you see other parameters that would require kwargs for the sampler? If not I would go with the first option.

Reply:
Okay, will do. Currently I do not see a concrete necessity. In the worst case, I guess the passed `sampler` would have to come with its `kwargs` fixed :)

Reply:
Done. I coded it up so that `sampler_seed` is passed to `sampler` for each of the `num_directions` iterations. I think for testing this makes sense, but it will basically destroy the stochastic averaging idea for `num_directions>1`, do we want that?
Alternatively, one could also just use `sampler_seed` in `_rademacher_sampler` if its the first direction, i.e. `idx_rep=0`. Let me know what you think! :)